### PR TITLE
feat(logging): Record user agent string

### PR DESCRIPTION
**Pull request by khatruong2009**
*Issue #, if available:*

*Description of changes:*
- Wraps customer invoked flush and batch send calls in identifyCall method allowing us to measure how many developer accounts are using it.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/amplify_core/lib/amplify_core.dart [hunk 1]
```diff
@@ -28,6 +30,7 @@ export 'src/config/notifications/notifications_config.dart';
 export 'src/config/storage/storage_config.dart';
 
 /// HTTP
+export 'src/http/amplify_category_method.dart';
```
Comment (NikaHsn) [inline packages/amplify_core/lib/amplify_core.dart:33]: issue: amplify_category_method.dart is an internal library so instead of exporting it from here use ```// ignore: invalid_use_of_internal_member``` in cloudwatch_logger_plugin.dart
```
// ignore: invalid_use_of_internal_member
import 'package:amplify_core/src/http/amplify_category_method.dart';
```

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart [hunk 2]
```diff
@@ -126,65 +126,70 @@ class CloudWatchLoggerPlugin extends AWSLoggerPlugin
...
-      final batchStream = _getLogBatchesToSync();
-      await for (final (logs, events) in batchStream) {
-        _TooNewLogEventException? tooNewException;
-        while (logs.isNotEmpty && events.isNotEmpty) {
-          final rejectedLogEventsInfo =
-              (await _sendToCloudWatch(events)).rejectedLogEventsInfo;
-          if (rejectedLogEventsInfo == null) {
-            await _logStore.deleteItems(logs);
-            break;
+    await identifyCall(LoggingCategoryMethod.batchSend, () async {
```
Comment (NikaHsn) [inline packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:129]: issue: this is called by flush api, timer sync and logStore full sync. instead of wrapping this we should use identifyCall where we call `_startSyncingIfNotInProgress` like:

```
_timer = pluginConfig.flushIntervalInSeconds > 0
        ? StoppableTimer(
            duration: Duration(seconds: pluginConfig.flushIntervalInSeconds),
            callback: identifyCall( LoggingCategoryMethod.batchSend,
                () => _startSyncingIfNotInProgress,),
            onError: _onTimerError,
          )
```

```
if (isLogStoreFull && _shouldSyncOnFullLogStore()) {
      await identifyCall( LoggingCategoryMethod.batchSend,
          _startSyncingIfNotInProgress);
    }
```

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart [hunk 3]
```diff
@@ -79,7 +81,10 @@ class CloudWatchLoggerPlugin extends AWSLoggerPlugin
     _timer = pluginConfig.flushInterval > Duration.zero
         ? StoppableTimer(
             duration: pluginConfig.flushInterval,
-            callback: _startSyncingIfNotInProgress,
+            callback: identifyCall(
+              LoggingCategoryMethod.flush,
```
Comment (NikaHsn) [inline packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:85]:
```suggestion
              LoggingCategoryMethod.batchSend,
```

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart [hunk 4]
```diff
@@ -293,7 +301,11 @@ class CloudWatchLoggerPlugin extends AWSLoggerPlugin
     );
 
     if (isLogStoreFull && _shouldSyncOnFullLogStore()) {
-      await _startSyncingIfNotInProgress();
+      await identifyCall(
+        LoggingCategoryMethod.flush,
```
Comment (NikaHsn) [inline packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:305]:
```suggestion
        LoggingCategoryMethod.batchSend,
```

**Review by NikaHsn — APPROVED** [review]

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart [hunk 5]
```diff
@@ -128,11 +133,14 @@ class CloudWatchLoggerPlugin extends AWSLoggerPlugin
   Future<void> _startSyncingIfNotInProgress() async {
     Future<void> startSyncing() async {
       final batchStream = _getLogBatchesToSync();
+
```
Comment (fjnoyp) [inline packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:136]: Why did you add these spaces in?  While some might make sense, having a new line in between a comment and the line it's commenting seems a bit strange: 

// set logs to end before the index.

logs.removeRange(tooNewStartIndex, events.length);

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart [hunk 6]
```diff
@@ -169,19 +183,20 @@ class CloudWatchLoggerPlugin extends AWSLoggerPlugin
...
+        nextRetry = DateTime.fromMillisecondsSinceEpoch(
+          e.timeInMillisecondsSinceEpoch,
+        ).add(_minusMaxLogEventTimeInFuture);
       } on Exception catch (e) {
         logger.error('Failed to sync logs to CloudWatch.', e);
       } finally {
-        _handleFullLogStoreAfterSync(
-          retryTime: nextRetry,
-        );
+        _handleFullLogStoreAfterSync(retryTime: nextRetry);
```
Comment (fjnoyp) [inline packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:198]: Do you have autolint setup?  I'm a bit confused by the inconsistent changes here. 

In the above line, you added a "," to a single parameter method call, but then below you remove a "," in a single parameter method call.

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart [hunk 7]
```diff
@@ -293,7 +308,11 @@ class CloudWatchLoggerPlugin extends AWSLoggerPlugin
     );
 
     if (isLogStoreFull && _shouldSyncOnFullLogStore()) {
-      await _startSyncingIfNotInProgress();
+      await identifyCall(
+        LoggingCategoryMethod.batchSend,
+        _startSyncingIfNotInProgress,
+      );
+      // await _startSyncingIfNotInProgress();
```
Comment (fjnoyp) [inline packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:315]: Remove

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart [hunk 8]
```diff
@@ -5,6 +5,8 @@ import 'dart:async';
 import 'dart:math';
 
 import 'package:amplify_core/amplify_core.dart';
+// ignore: invalid_use_of_internal_member, implementation_imports
```
Comment (fjnoyp) [inline packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:8]: Looking through our repo I don't see any precedent for ignoring these warnings.  What is the reason we need to ignore this warning here?  "invalid_use_of_internal_member"

File: packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart (hunk 5 above)
Comment (fjnoyp) [reply packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart:136]: Have you run `dart analyze` to remove any warnings beforehand as well?  Perhaps it's not necessary since this is a feature branch, but might be good to start running those to get ahead of the warnings (if there are any)

**Review by fjnoyp — CHANGES_REQUESTED** [review]
Added some syntax/nit comments otherwise looks good.

**Review by fjnoyp — APPROVED** [review]
